### ISSUE TITLE
How to deal with breaking changes on platform ? [OpenBSD]

### ISSUE THREAD

**Issue by semarie** [rust-lang/libc#570]
I open an issue on libc because it is here the problems will start to show up. Depending the solution or the way to deal with, modifications could occurs in rustc repository too.

At OpenBSD, we don't care about breaking API/ABI between releases. Once a release is done, the API/ABI is stable, but there is no guarantee that it will be compatible with the next release.

Currently, in the upcoming 6.2 version of OpenBSD (6.1-current), there is a breaking change that will affect libc : [si_addr should be of type void *, not char * (caddr_t)](http://cvsweb.openbsd.org/cgi-bin/cvsweb/src/sys/sys/siginfo.h.diff?r1=1.11&r2=1.12). Here the current definition in [libc](https://github.com/rust-lang/libc/blob/master/src/unix/bsd/netbsdlike/openbsdlike/mod.rs#L110).

Under OpenBSD, we deal with ABI under LLVM by using a triple like: `amd64-unknown-openbsd6.1`. For Rust, instead we use an unversioned platform, resulting all OpenBSD versions to define the same ABI (which isn't properly right).

Do you think it possible to switch from `*-unknown-openbsd` to `*-unknown-openbsd6.0`, `*-unknown-openbsd6.1`, ... without having to duplicate all code in libc for each target ? and without having to add a new target in rustc for each OpenBSD release ?

Any others ideas on the way to deal with it ?

**Comment by alexcrichton** [comment]
Unfortunately I don't really know how we'd handle this, I just figured that platforms wouldn't do this. 

If this happens a lot we'll just need to document what's wrong and stop adding new bindings, it'll be up to crates to implement version compatibility.

**Comment by semarie** [comment]
I think it isn't just a "version compatibility" issue. My purpose isn't to have a compatibility layer for missing/removed functions or types.

The problem is OpenBSD triple is versioned, meaning that API/ABI of one version could be different from another version.

I checked some others system (running `llvm-config --host-target` to see if the triple is versioned or not), and it seems it is a common situation in not-Linux world:
  - x86_64-apple-darwin16.0.0
  - x86_64-unknown-freebsd12.0
  - x86_64-unknown-freebsd11.0
  - i386-unknown-openbsd5.8
  - x86_64-unknown-netbsd7.99

I also checked in LLVM source tree: the OS version is a part of the triple definition. 
see [getOSVersion()](https://github.com/llvm-mirror/llvm/blob/master/include/llvm/ADT/Triple.h#L310) in `include/llvm/ADT/Triple.h`.

Maybe a concept is missing in Rust ? If `target_os_version` would be available, it would solve the issue: parts that are only defined in some OS version could be isolated from another OS version.

I don't think it is a problem only on OpenBSD. Any OS using OS-Version could be hitted. OpenBSD exposes it because we heavy use the ability to *not* be API/ABI compatible (it is a way to be able remove old stuff that deserve security).

**Comment by alexcrichton** [comment]
Yeah there's no concept of a versioned target in rustc right now, and we're unfortunately not really capable of doing so right now.

Our only recourse is basically to take the subset which currently works across all revisions, put that in libc, and then otherwise let downstream crates bind versions that change over time.

**Comment by semarie** [comment]
I hope you are kidding: you are asking to remove `siginfo_t` type for OpenBSD from libc and so to break stack_overflow detection for OpenBSD (libstd relies on it). And even if we can drop stack_overflow detection, it doesn't resolv the intrinsic problem.

So I am looking to extend Target to include os-version information in the target specification.

**Comment by alexcrichton** [comment]
Well, I'm not really kidding. If we feel we must fix this then we currently have no choice but to not add the bindings. If we don't want to do that then the fix must go elsewhere. I don't know the best way to fix this, just spitballing.

**Comment by asomers** [comment]
This isn't just a problem for OpenBSD.  FreeBSD 12, when it comes out, will change a number of important types, like `ino_t` and `struct stat`.  If libc's policy is to only bind the greatest common denominator between versions, then overtime it will shrink into irrelevance.  Such a policy really just kicks the version compatibility can down the road.

Would it be possible to generate bindings dynamically at build time?  When writing Ruby bindings, I've always preferred that approach to FFI.  If not, then I think libc needs a way to distinguish between OS versions, just as it currently distinguishes between OSes.

**Comment by alexcrichton** [comment]
I would personally be afraid of generating bindings at compile time. It just pushes the problem to consumers without giving them tools to deal with it.

I do think that this sounds like this needs a way for libc to distinguish between OS versions, but Rust currently has no tool for doing so really.

**Comment by asomers** [comment]
The problem just got worse.  Linux 4.11, released today, added a new system call: `statx`.  Until libc learns to understand versions, it can't add support for `statx`.  I really thing that cargo needs some sort of configure step analagous to autoconf's configure.
https://www.phoronix.com/scan.php?page=news_item&px=Linux-4.11-Statx-System-Call

**Comment by alexcrichton** [comment]
@asomers that's not quite true though, we can add bindings at any time. Rust supports Linux 2.6.18+ and there are a huge number of syscalls bound in libc not present in 2.6.18. It's up to crate authors to pick and choose apis for platform compatibility appropriately.

**Comment by asomers** [comment]
@alexcrichton I guess I was wrong about how libc's CI tests worked.  Are you saying that libc's tests do not flag symbols defined in FFI but not present in the system's headers?  If that's true, then a Rust program trying to use `statx` on Linux <= 4.10 would build but get `ENOSYS` at runtime, right?  That's better than what happens if a Rust program tries to use `aio_waitcomplete` on FreeBSD 10, where the FFI binding would actually be wrong.  But Linux is not immune from changing syscalls, either.  The first example I could find was `utimensat`.  Its signature changed in 2010, well after 2.6.18 was released.  Any Rust program using libc will try to use the new version of `utimensat`, even when built for older systems.

Would you consider dynamically generating the bindings for select functions, even if most functions have static bindings?  Right now, I don't see any way at all for consumers to deal with the versioning problem.

**Comment by alexcrichton** [comment]
No, to be clear:

* The `libc` crate is basically just a header file.
* The `libc` crate is automatically verified on many platforms, but we have exceptions in `libc-test/build.rs`. It's not guaranteed that the verification passes on every instantiation of a platform.
* If you use a function from `libc` you're referencing a symbol.
* If that symbol doesn't actually exist on your system, you'll get a linker error.

Programs using `statx` will likely get a linker error and will then have to deal with that appropriately.

I would like to avoid dynamically generating the bindings, as that's typically not the actual solution to this problem. It makes cross compilation (even just across OS versions) much more difficult

**Comment by asomers** [comment]
Cross-compilation could be solved by overriding the build script's platform detection.  For example, on FreeBSD there's basically only one symbol that a build script would need to detect: `__FreeBSD_version`.  When cross-compiling, cargo could set that in an environment variable and the build script wouldn't try to detect it from the system headers.

But it sounds like you might have something else in mind when you say it's "not the actual solution to this problem".  Do you?  What is the "actual solution", @alexcrichton ?

**Comment by alexcrichton** [comment]
To me the "actual solution" is precisely what we're doing right now. We list a bunch of symbols and authors need to be vigilant about which ones they use. This does not solve the use case of OpenBSD, however, if there are ABI breaking changes. We may need more than one solution but to me there are too many downsides to dynamically generating an API on Linux at least.

**Comment by asomers** [comment]
Not only does it not solve OpenBSD's use case; it doesn't solve the use case where operating systems make changes that _don't_ break the ABI.  Both FreeBSD and Linux occasionally change syscalls and provide backwards compatible syscalls with the old signature and syscall number but a new name.  For example, FreeBSD 8's "compat7.shmctl" syscall is identical to FreeBSD 7's "shmctl".  Similarly, operating systems make changes to system libraries and provide backwards compatibility by bumping the SHLIB version and providing the old libraries as optional packages.

Currently libc handles neither of these cases.  Either the libc binding tracks the new function's signature, which breaks Rust programs at runtime on older versions, or the libc binding stays with the old signature, which breaks Rust programs at runtime on newer versions.  It's simply not possible for the current libc to compile correctly on multiple versions of an operating system.  Your previously suggested solution is to simply remove a binding whenever the OS changes it.  But that would break any crates that use the binding, violate semver, and still result in runtime failures for crates that use the old libc but were built on a new OS.

You suggest that libc's consumers should be responsible for versioning issues, but I don't think that's possible.  Let's take `stat(2)`, for example, which will likely change in FreeBSD 12.  Suppose that when FreeBSD 12 is released, somebody tries to compile the `nix` crate on it.  The linker will be satisfied that `stat` is present in libc, but the signature will be totally wrong, so `nix` will fail at runtime.  Cargo won't produce any kind of warning.  If I understand you correctly, you suggest that `stat` should at this point be removed from libc.  But that won't fix `nix` until somebody updates that crate's dependencies, and even then it will only change a runtime failure into a compile time failure.  Must the `nix` developer then write a build script that checks `__FreeBSD_version` and reimplement all of `stat`'s FFI bindings for FreeBSD 12?  That would finally fix the problem.  But according to crates.io, libc has 1268 dependent crates, and _all_ of them would have to independently write the same build script and add the same FFI bindings for `stat` on FreeBSD 12.

Alternatively, libc could assume that all operating systems provide backwards but not forwards compatibility (sorry OpenBSD).  Then it could pick a minimum supported version, and always link against that version's shared libraries,  Currently Cargo doesn't provide a mechanism to specify an exact shared library version to link against, but that could be added.  This would fix all of the runtime failures, but at substantial cost: dependent crates would lack access to new OS features, and both developers and users would have to install the compat library packages.  Not only would new features that change APIs be unavailable, but the shared library lock would mean that entirely new functions would be unavailable as well, unlike the current situation where using newly added functions will generate link failures when building on an old OS.

In either case, developers will likely fork libc to update their favorite bindings, resulting in a Balkanization of libc and dependent crates that don't support older OS versions.

I understand that cross-compilation is a really cool feature, but I fear that you're underestimating the severity of this problem.  Have you looked into how embedded cross development toolchains work?  AFAIK the host system requires full headers for the target.  Maybe Rust needs to do the same.

**Comment by alexcrichton** [comment]
@asomers if you've got a proposal of what to do I'd recommend writing up an RFC, with so many dependencies changes such as what I think you're proposing can't be taken lightly.

**Comment by semarie** [comment]
@alexcrichton I pushed a WIP branch on my github repository. I hope code will be more explicity than my explaination about what I called having support for OS version.

Tree is at https://github.com/semarie/rust/tree/target-os-version . Please note my code isn't working for now.

Basically, it is:
  - extending Target to embedding a (possibility empty) os-version string
  - exposing the string as `target_os_version` symbol (in the same way than `target_os`)

It would be possible to do have conditionnal code against OS version (OpenBSD 6.1 or OpenBSD 6.0), in the same way we have conditionnal code against OS name (OpenBSD or FreeBSD).

**Comment by asomers** [comment]
Good work @semarie.  BTW, I've been studying ELF symbol versioning and I think it would be possible to fix libc without modifying Rust itself.  Basically, libc would need to grow a bunch of feature flags like "freebsd11+", "freebsd10+", etc, meaning "build code that will work on FreeBSD 11 or greater" and "build code that will work on FreeBSD 10 or greater".  Of course, those flags could be conditionalized so they won't appear on other platforms.  Then, for every symbol that differs between FreeBSD versions, libc will bind a different version depending on which feature flags are set.  The `link_name` attribute will encode the specific ELF symbol version number used on the oldest OS version chosen.  I don't have code yet, but I think this approach will work for FreeBSD and Linux.  Does OpenBSD use ELF binaries or is it still using a.out?

Also, I've found several functions in glibc with multiple versions.  Linux is not immune from this problem.

**Comment by semarie** [comment]
@asomers OpenBSD uses ELF on all platforms. but using ELF symbol versioning doesn't help for breaking changes if the OS doesn't use symbol versioning.

**Comment by alexcrichton** [comment]
@semarie I'd personally probably reocmmend writing an RFC before sending that as a PR, I'm sure many others would have comments as well!

**Comment by raphaelcohn** [comment]
This has hit me too - in particular, with changes in Mac OS X major versions. However, a good solution probably isn't to use a version number in the target triple, as there's a distinction to be drawn between libc version and OS version; they do not necessarily go in lockstep. A classic example might be changes to Linux's uapi headers, which don't yet line up with changes in musl, say.

This problem is a general one: changes in third party (usually C) library APIs that are incompatible. It needs a good solution within Rust. It's a problem that's heavily compounded by set ups that use dynamic libraries (something I've come to see as more trouble than they're worth for secure or robust systems outside of the desktop. In practice, it's a far too difficult for most sysadmins to assess whether a security fix to a dynamic library affects more than on running program, and so they just go for the nuclear option of a reboot). Using `autoconf` like tests or dynamic bindings at runtime is probably the wrong way to solve this generally. Such approaches require too much of the system they are on (execute permissions, existent of compilation-associated tools, headers, etc). They are also deeply unfriendly to security audits and locked down systems (eg those built entirely from source). `autoconf` in particular makes the classic mistake that the build system is similar to deployment; it's always been an absolute beast to get things to cross-compiles with it repeatedly, robustly and consistently. Too many things (eg time-of-date, location of shell interpreter, absolute sysroot paths, etc) creep into the deployed solution.

(Semantic version does absolutely nothing to solve this; in fact, semantic versioning is a deeply broken concept that's become popular recently. In practice, either a version is compatible or it isn't; semantic versioning is just the upstream's author's assessment. One man's inconsequential security version or minor change is another's nightmare. In practice, with large system set ups and deployments, I always encourage dev teams to think of only two kinds of version: likely-to-be-compatible security fix, and incompatible. Everything incompatible needs to go through the full test cycle before deployment. Security fixes can bypass that if urgent; risk vs reward and all that).

**Comment by comex** [comment]
Er, is the `siginfo.h` change in question actually ABI-breaking?  `char *` and `void *` should have the same memory representation, so I'd expect that change to break the API (for newly compiled C code) but not the ABI.

Though it seems there's a more general problem here.

**Comment by semarie** [comment]
@comex yes, the change `char *` to `void *` is just an API break regarding OpenBSD. But it is an uncommitable change in crate libc without major version bump.

I started a discussion on [internals](https://internals.rust-lang.org/t/pre-rfc-target-extension-dealing-with-breaking-changes-at-os-level/5289), and I am working to submit a RFC.

**Comment by Rufflewind** [comment]
Related: https://github.com/rust-lang/rust/issues/42681

> fs::metadata() crashes on FreeBSD 12 due to layout change in stat.h

**Comment by mattmacy** [comment]
https://github.com/rust-lang/libc/issues/775

I don't quite understand how Rust initially missed out on OS and ABI versioning quite so badly - deciding to  assuming that structures and types are immutable over time or removing key structures from libc altogether. Nonetheless, Rust can conditionally compile based on configuration values, what is stopping this?

**Comment by semarie** [comment]
@mattmacy my understanding of the problem is:
- build time configuration adds more complexity for crosscompiling (you need to target a particular OS ABI)
- crosscompiling is used a lot in Rust infrastructure (for testing, for rustup...) So using OS ABI would mean a potentially high number of new targets to check, and binaries to produce, resulting an infrastructure more complex to maintain (it needs to scale)
- it is only a problem for BSDs, and they are not a high priority

**Comment by asomers** [comment]
@semarie It's not just a BSD problem.  Even glibc occasionally bumps the ELF symbol version of a function.  Also, the version proliferation problem isn't quite as bad as you're thinking.  Using feature flags and symbol versioning, we could build libc for targets like "freebsd10+", "freebsd11+", and "freebsd12+".  A project could get pretty good test coverage just by using "freebsd10+".

**Comment by mattmacy** [comment]
@asomers  @semarie This is an artifact of myopic developers who have nothing more than a superficial understanding of the Linux ecosystem post-2010 - Linux is the post-M$ monoculture. This is going to be a pervasive issue over the longer term and thus runs counter to Rust's mission statement as a general system's language. This isn't rocket science and I hope we don't end up needing to go down the Go "replace all the host libs" route.

**Comment by mattmacy** [comment]
@kev009 - Exactly. A big selling point of Rust is its efforts to eliminate UBs. In light of that, no one can really claim to be "cross platform" until this issue is formally addressed.

**Comment by semarie** [comment]
@asomers I don't understand in depth ELF symbol version of a function, and how particulary how glibc uses it. So I couldn't comment on that. But Rust language has syntax to bind a particular extern function (in Rust) to a specific symbol (in library).

What I know is some OS (and particulary OpenBSD) has different policy regarding compatibility between OS versions. In particular for OpenBSD, ABI could change between OS versions, so it is expected to a compiler to target a *specific* OS version. OpenBSD does it only if required (we don't do it just for pleasure), but it happens.

The current case with FreeBSD is interesting. First it is FreeBSD and not OpenBSD. The users  base is more important, and for Rust the platform is tier-2 (OpenBSD is tier-3). The kind of change (passing underline `ino_t` struction from 32 bits to 64 bits) is problematic because it affects structure ABI and not only function ABI.

FreeBSD provides a thin compatibility layer for helping a binary compiled for FreeBSD 11 to run on FreeBSD 12, but the intent is to help migration, not to ensure a full compatibility (64 bits inode *isn't* representable on 32 bits structure ; the default behaviour on FreeBSD 12 is truncation). I assume OpenBSD doesn't provide such mecanism specifically for avoiding third-party to abuse them (like only using the compat layer for simplicity), and keep a clean and tested base (compat layer isn't a standard code, so it is less used and less tested).

**Comment by comex** [comment]
The entire point of ELF symbol versioning is to *prevent* breaking ABI backwards compatibility, by allowing programs linked against an old version of the library to continue to use the old definition.  glibc, as a well-behaved library, would not intentionally break ABI backwards compatibility without bumping the version in the soname, i.e. libc.so.6 to libc.so.7.  Given that the transition from 5 to 6 happened in 1997, I don't expect that to happen anytime soon.

macOS also tries to preserve ABI backwards compatibility, and I'd like to hear more about the issues @raphaelcohn was experiencing.  (It too migrated from 32-bit to 64-bit inodes once, but the compatibility layer remains present to this day.)

What I suspect is that both people are actually talking about *forwards* compatibility breakage, not backwards - that is, binaries compiled against new headers won't run on old OSes, which can easily happen even if the code itself doesn't use any functionality specific to the new OS.  That's an important issue, but it's different from backwards compatibility breakage, as it's more amenable to solutions that don't require maintaining totally separate builds for the two versions, e.g. using runtime detection.

**Comment by semarie** [comment]
Here a resume **for OpenBSD** : the rule is compatibility isn't guarantee neither for backwards or forwards between OS releases (all releases are considered as major release versions).

But in general, it is possible to run a binary from version N-1 on version N, but it is only a facility provided if possible (without to much effort for OpenBSD developers). Older things are just [removed](https://www.openbsd.org/papers/pruning.html) (it helps to keep the system in a well tested and audited state, without keeping old crafted code paths).

Once released, the specific OS version has ABI/API stability guarantee. Only security and reliability changes will be commited on the branch: they are published [errata](https://www.openbsd.org/errata61.html).

OpenBSD doesn't use ELF symbol versioning. But libraries (system and packaged third-party) have a strict major/minor bump policy (we have currently a `libc.so.90.0` in 6.2-beta, and OpenBSD 6.1 has `libc.so.89.3`).

**Comment by asomers** [comment]
@comex you are confusing ELF symbol version with soname versioning.  Libraries bump their sonames due to major changes, like ncurses did two years ago and glibc did in 1997.  But ELF symbol versioning is used to maintain backwards compatibility while making minor changes to the library.  Glibc is very conservative in its use of ELF symbol versioning, but it _does_ occasionally bump a symbol version.  Look through its sources and you'll find a few examples.

Rust's difficulty comes from its extensive use of FFI and cross-compiling, which cause it to ignore all header files.  The problem isn't backwards or forwards compatibility per-se.  Rather, it's that libc simply cannot express bindings for more than one version of a library.  Right now, it simply isn't possible to build a Rust program that uses libc and will run on FreeBSD 12.  It doesn't matter what host you build it on.  And if libc ever updates its bindings to the FreeBSD 12 ones, then it will no longer be possible to build a Rust program that will work on FreeBSD 11.

**Comment by mattmacy** [comment]
@asomers As an old timer I can tell you that Linux distros never used symbol versioning with great success and thus shared libraries were viewed as broken by designl. At my work places they would turn off automatic RHEL security updates because it frequently crippled all their systems with incompatible symbol resolution. Moving to a frozen ABI is a (healthy) response to their inability to make versioning work. The Rust developers assumption that everything looks like Linux has led them to overlook the fact that the source contract is not the same as the ABI contract. Building for a given version of the OS assumes that you're using the system headers from that version. FreeBSD will happily run *dynamically linked* binaries at least as far back as 4.x if not much further. However, if your building on version X you're expected to use headers for version X, not whatever version you originally chose to copy the definitions from.

**Comment by comex** [comment]
> @comex you are confusing ELF symbol version with soname versioning.

I understand the difference, and I think I stated it correctly in the first paragraph of my last comment, but perhaps my wording was unclear.

My point was that even if glibc occasionally bumps a symbol version, that is not a backwards compatibility break (for old binaries), unlike the situation with - at least - OpenBSD.  According to @mattmacy's latest comment, FreeBSD actually doesn't break old binaries either; in that case #42681 is 'just' a mismatch between Rust bindings and the version of the system present in the build sysroot, *not* the version of the system present at runtime.  I think FreeBSD implements this using ELF symbol versioning.  If so, Rust could hypothetically choose to compile 11-compatible binaries regardless of whether the sysroot is version 11 or 12, by explicitly specifying a symbol version in the object file rather than leaving it to the linker; it could then add runtime detection to use the FreeBSD 12 symbols if available.  I could be wrong about how it works, and in any case I'm not sure whether such a compatibility measure would be desirable.  The alternative would be, as @asomers suggested, creating targets like freebsd11+ and freebsd12+, corresponding to the sysroot version.

But in general, I perceive a few differences between backwards-compatibility ABI breaks and forwards-compatibility breaks:

- With forwards-compatibility breaks, the changes are *probably* not so far-reaching as to require excessive duplication in bindings that support both the old and the new version.  After all, the library's own codebase must contain similar duplication.  Admittedly, this is a heuristic, and "excessive" is in the eye of the beholder.  As one example, on my Mac, I count 38 libc functions (quite a lot) that have duplicate 32-bit-inode and 64-bit-inode versions, and at least two structs (`struct stat`, `struct statfs`).  On the other hand, backwards compatibility breaks have no limits on how much they can change, and no reason to care about the difficulty of supporting both the old and new ABI in one binary.

- Ditto in time rather than space: on systems designed for ABI forwards compatibility, subtle changes like adding new symbol versions probably won't occur too *frequently*, because it's a pain to deal with them on the implementation side, and the old versions have to be kept around 'forever' (well, at least for some time).  On a system that breaks ABI backwards compatibility every version (but preserves API compatibility), someone might decide to fiddle with a struct's layout for 5 versions in a row (because why not?), and an omni-compatible binary would have to contain 5 alternate implementations for everything that uses the struct.  On systems that preserve backwards compatibility, in most cases, you're unlikely to see more than two versions of any given symbol.

- With backwards-compatibility breaks, users do not expect the same binaries to run on both versions, and will generally replace all their binaries when upgrading.  Although in most cases it would be *possible* for Rust to produce binaries that run on both via runtime detection, it may not be that useful, since it wouldn't fit into common user workflows.  By contrast, for forwards-compatibility breaks, there's a preexisting concept of "X-or-later compatible" binaries, and a preexisting pattern whereby binary distributions of C programs can choose to maximize their compatibility by building against the oldest possible version - a pattern which Rust can directly emulate, and improve on by adding runtime detection.  But this depends on cultural factors too.  On, say, macOS, targeting old OS versions is a given, and you can do so while compiling against the latest SDK just by passing an appropriate value to -mmacosx-version-min; the compiler and system headers work together to ensure the output binary is compatible with the specified version.  On FreeBSD, from what people have written in this thread, it seems like binary compatibility is seen more as a stopgap measure.  There's an open-source vs. proprietary angle here, of course...

- With forwards compatibility breaks, you can generally query at runtime whether the old and/or new symbols are available, just looking at the symbols themselves.  With backwards compatibility breaks, it may be necessary to resort to indirect measures (like the OS version), which are less reliable.  As a semi-related example, rustc has a C++ helper for LLVM features that aren't exposed via the stable C API, only the unstable C++ API, and it contains a ton of #if blocks for 'is LLVM version greater than X'.  But at one point I tried building against Xcode's LLVM, which was a branch of an older trunk version with newer changes cherry-picked on top.  API-wise, it mostly corresponded to an LLVM version different from the one it declared, which confused the heck out of Rust's helper - but changing the declared version wasn't enough, because it wasn't *fully* consistent with any one version.  I imagine you could have similar issues with OS ABIs :)

- Also, 'always use the old symbol' is always available as a starting point.  It's usually not a good ending point; chances are the ABI break happened for a reason.  If Rust decided to always target 32-bit inodes on FreeBSD, the resulting binaries would 'work' on all OS versions, but they'd break at runtime when encountering a file whose inode actually can't fit into 32 bits.  But there may be cases where that isn't really necessary.

- As a special case of the above: a system designed for ABI forward compatibility may have *API* changes that from the ABI's perspective are no change at all, like the `void *` vs. `char *` thing earlier in the thread.  In this case, Rust should seriously consider always using one or the other.

All that said, even when increased compatibility compared to C is possible, its benefits have to be weighed against fundamental drawbacks.  One is the binary bloat of runtime detection.  Another is simply the fact of Rust acting *differently* from C.  One of Rust's biggest selling points is that it feels like a native citizen of the OS, a companion to C; it doesn't live in its own world with its own VM or toolchain.  Thus, users may justifiably expect the `libc` crate to work the same way, both API- and ABI-wise, as native C on the same platform.

Then again, Rust does already have its own build system and its own conception of 'target'.  In addition, whereas C compilers are generally shipped as operating system components, with each OS vendor responsible for getting everything working on their OS, Rust has rust-lang.org binaries as a major distribution mechanism for the compiler, so it's more desirable to limit the number of separate targets that have to be built and maintained.

**Comment by mattmacy** [comment]
@comex Can you distill the implications of your last post? 

Yes, Rust has it's own notions of target and outside of Linux and Windows that notion is _wrong_. I just want to be able to use Rust across different 3-tuples (OS/ABI/arch) as opposed to the current 2-tuple (OS/arch) with the ABI set to wherever they happened to copy the header values from.

**Comment by mattmacy** [comment]
Although I managed to get jsonrpc_http_server bits to work yesterday by patching some of the dependent ports and then using [patch.crates-io] overrides, today I tried building netmap_sys with user libs which failed in the gcc crate, presumably because of something in libstd. I tried installing the FreeBSD port, but the cargo in that version wouldn't let me override libc (if that's something only available in nightlies you really need to make it standard) so although it compiled the binary didn't actually work. I've lost too much time on this to consider Rust a viable option for any near-term deadlines and have concluded that Rust really isn't cross-platform outside of tier 1 targets right now.

**Comment by comex** [comment]
@mattmacy To quote http://doc.crates.io/manifest.html:

> Note that the [patch] feature will first become available in Rust 1.21, set to be released on 2017-10-12.

In the mean time you could use [replace].

**Comment by mattmacy** [comment]
This is why this oversight really perplexes me. Having the version be an explicit part of the OS target is not hidden and is visible to anyone who has looked at a configure script for any cross-platform OSS program:

```
 | |                     \-+- 69334 mmacy make DIRPRFX=lib/clang/ all
 | |                       \-+= 69335 mmacy sh -e
 | |                         \-+- 69336 mmacy make all DIRPRFX=lib/clang/libllvm/
 | |                           |-+= 71356 mmacy sh -ev
 | |                           | \-+- 71357 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71358 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71499 mmacy sh -ev
 | |                           | \-+- 71500 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71501 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71769 mmacy sh -ev
 | |                           | \-+- 71770 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71771 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71772 mmacy sh -ev
 | |                           | \-+- 71773 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71774 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71778 mmacy sh -ev
 | |                           | \-+- 71779 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71780 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71781 mmacy sh -ev
 | |                           | \-+- 71782 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71783 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           |-+= 71784 mmacy sh -ev
 | |                           | \-+- 71785 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                           |   \--- 71786 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
 | |                           \-+= 71787 mmacy sh -ev
 | |                             \-+- 71788 mmacy c++ -O2 -pipe -I/home/mmacy/devel/build/usr/home/mmacy/devel
 | |                               \--- 71789 mmacy /usr/bin/c++ -cc1 -triple x86_64-unknown-freebsd11.1 -emit
```

**Comment by kev009** [comment]
@comex I think you're starting down the right track now.  The current rust libc is trying to do something that is not really tractable, an impedance mismatch with C language calling and linkage conventions.  In effect:  lazy binding with a static set of bindings.  That feels like something you might do in a dynamic language, and there you might not care much about correctness or performance and can further do runtime introspection or generation to smooth out the issues.  So I do feel strongly the current Rust behavior here is undefined behavior on all platforms and think it is unfair to characterize it as a BSD issue.  It happens to work widely for now, but it is a major design flaw that just happen to be seen on BSD right now and will fester as rust wants to grow.

Let's step back to buildeng history for a bit.  If I were a major ISV, I want to support some forward range of OS versions.  We'll chose Red Hat EL.  So I build on EL5, and I build my app on there and it will work for probably a very long string of future versions perhaps with compat packages that maintain an old sover since sover ought to be bumped for ABI changes.  You can substitute Red Hat EL for Windows or HP-UX or anything else, OS makers usually strive hard to keep that kind of forward compat.  There is one exception, OpenBSD, where forward compat is not guaranteed at all, and binaries are expected to match the running major.minor.  But OpenBSD's choices aren't really an issue for C language authors if you understand the root issue, they just choose no amount of forward compat.  You build on the oldest platform you wish to support for all OSes.  As an implementation detail this could be in chroots, jails, containers or just plain path, compiler and linker flag gymnastics for cross building.  That's how things have been done for a long time.

Now, there is a second and much less reliable case for building with backwards compat.  In a big IDE this might be a hidden build process which looks like the above path/flag munging to target the compiler to the right runtime and linkage.  In something like MSVCxx that might be by bundling a runtime environment with the installer.  I'm not sure rust should bother with this at the moment, but it's not something I've put a great deal of thought into.

Another thing I haven't put a great deal of thought into is direct usage of syscalls by the language or libraries.  Things discussed like soname and sover go out the window with that.

**Comment by jakllsch** [comment]
I've run into this issue while trying to ascertain the correctness of the libc crate on NetBSD 7.x and 8.x.  The current situation has the libc crate using some combination of 7.x and 8.x ABI/APIs, while the official distribution artifacts for x86_64-unknown-netbsd currently target 7.x.  Both NetBSD 7.x and 8.x are currently-supported release serieses.

I really want a way to express these and future ABIs from a single version of the libc crate, that can target a specific OS kernel-libc version at compile time.  Targeting the current oldest supported NetBSD release series might be a reasonable way forward if that is impractical, as there is libc-level and kernel-level backward compatibily across many major versions for old binaries.

**Comment by gnzlbg** [comment]
@alexcrichton 

> Our only recourse is basically to take the subset which currently works across all revisions, put that in libc,

So what do we do with the stuff that's in libc but broken? You mention that if a symbol changes, you get a linker error, but what happens if the symbol does not change, and the API does ? What do we do with `const` that have incorrect values ?

I see at least two actionable alternatives: we deprecate broken stuff or we update it to some platform version. It appears that, currently, at least for constants, we have chosen the path of updating things. If that's the path that libc is going to take, we could also take it for functions and types as well (e.g. by leaving the old one in place, deprecating it, and adding a new one, e.g., called `foo2` - for functions we could link them both to the same symbol). Also, libc should be crystal clear about which exact version of each platform it "supports" and have a policy for updating (e.g. we always follow the latest version).

**Comment by alexcrichton** [comment]
@gnzlbg afaik all we can really do is just do our best as it comes up. I'm not sure we need to preemptively make policies here but rather we can see what policies we need later down the road if necessary.

**Comment by semarie** [comment]
I am still sad about that.

Please remember that currently FreeBSD 12 uses a compatibility layer for `ino_t` (FreeBSD 11 uses 32bits version, and FreeBSD 12 64bits) and silenciously truncate inode values if not representable on 32 bits. One day it will hurt, and it could be badly.

Refusing to accept the reality that OS introduces breaking changes from time to time, and enforcing that at libc level is bad.

**Comment by gnzlbg** [comment]
> Refusing to accept the reality that OS introduces breaking changes from time to time, and enforcing that at libc level is bad.

That OS introduce breaking changes is a fact, and AFAICT everybody agrees that this is a problem. 

I personally haven't heard a solution to this problem that did satisfy our constraints and did not have major downsides. Most proposed solutions have a high cost and the actual added value of solving this problem is unclear - this is a very tricky problem to solve correctly. 

For the FreeBSD case, the ""best"" solution might be to just bump the minimum supported FreeBSD version to FreeBSD 12, and if there is enough demand, to add a `...-freebsd11` target, or to provide an easy way to compile to such a target using `xargo`. But all of these changes are breaking, so they aren't great either =/

**Comment by asomers** [comment]
For FreeBSD, I think the easiest solution would be to add a default-on `freebsd11-compat` feature flag to libc.  When on, libc binds the FreeBSD 11 version of symbols like `struct stat`.  When off, it binds to the FreeBSD 12 version.  This wouldn't require introducing any new target triples.  However, that technique wouldn't help OpenBSD users.  OpenBSD pretty much needs a new target triple for each version, since it doesn't preserve any backwards compatibility.

**Comment by raphaelcohn** [comment]
@gnzlbg @asomers Whilst such techiques are likely to work, the downside is that a large number of 'legacy' triples accumulates over several years. Over a decade, a large number of the *BSD triples would effectively be obsolete, and hence challenging to maintain (The "matrix of pain" grows too large even for an experienced team).

One only has to look at GNU's disastrous `config.sub` / `config.guess` code to see what can happen. (They also show why target triples are exceedingly naive; at least Rust has gone down a much cleaner route by defining toolchain features with its JSON target files and dumped the idiocy that is feature testing).

The challenge for what to do is harder still when one uses C libraries that are NOT tied to a particular OS release, yet occasionally change to maintain compatible. A classic example is musl, which adds features with each release to improve its 'compatibility' with POSIX and other C libs, and every now and then also changes to accommodate a Linux kernel revision. (I've argued previously for separating core C library features from Linux headers, but the two are so tightly coupled in some views of the Linux world that it's impossible).

A better long-term approach may be to actually slim down libc to just a sub-set of POSIX, and then provide companion crates, eg for musl, glibc, ulibc, FreeBSD X.X, etc, which contain library-specific features. It may be that a companion crate might actually cover several very closely related libraries (eg musl + glibc + ulibc), but even that has difficulties. And even if there's a package per c-lib, there still needs to be a sensible way to handle c-lib versions that are moderately recent. I suppose eventually one just ends up with a lot of feature flags.

Of course, this gets challenging when also needing to consider Windows - but the vast majority of useful Windows functionality is outside of the library in any event.

**Comment by gnzlbg** [comment]
@raphaelcohn I agree that these approaches have downsides. I agree with you that: 

> A better long-term approach may be to actually slim down libc to just a sub-set of POSIX, and then provide companion crates, eg for musl, glibc, ulibc, FreeBSD X.X, etc, which contain library-specific features. It may be that a companion crate might actually cover several very closely related libraries (eg musl + glibc + ulibc),

And for `windows` the `winapi` crate already does this. For `macos` there are actually a couple of system libraries, and these are not exposed in one but in many crates. So each platform could have its own crate (e.g. `freebsd`, `openbsd`, etc.) with its own release schedule, and as long as the fundamental types (e.g. `c_int`) are defined somewhere else, all of these can interoperate. 

The downside of this approach is that "slim down `libc`" is a backwards incompatible change. We might be able to extract the parts of `libc` that `libstd` actually uses into its own crate, and make sure that `libstd` is as "version independent" as possible.

**Comment by ssokolow** [comment]
> A better long-term approach may be to actually slim down libc to just a sub-set of POSIX, and then provide companion crates, eg for musl, glibc, ulibc, FreeBSD X.X, etc, which contain library-specific features. It may be that a companion crate might actually cover several very closely related libraries (eg musl + glibc + ulibc), but even that has difficulties. And even if there's a package per c-lib, there still needs to be a sensible way to handle c-lib versions that are moderately recent. I suppose eventually one just ends up with a lot of feature flags.

What a coincidence. I don't see much on this topic for years and years and now, in the same month, not only you but also the Linux developers were discussing ideas in this vein:

* https://www.phoronix.com/scan.php?page=news_item&px=Linux-Propose-System-Wrap-Lib
* https://lkml.org/lkml/2018/11/10/470

**Comment by raphaelcohn** [comment]
@gnzblg I agree it would be a backwards incompatible change; it really depends how it is handled. At the moment, libc is 'stuck' in a 0.2 series of releases and so there is an opportunity - although it would require a major effort. Something like 6 months full time, I suspect, just to get a new, robust and agreed shape together.

I was thinking a little more over the weekend inspired by
>> with its own release schedule
and I think that would also allow a solution to version legacy and its associated maintenance. A new major release of a per-platform-like library would be free to drop support for an older version. Alignment with OS releases (or at least, with deprecation of long term support releases and the like) becomes quite possible, but historic code, because of cargo's excellent versioning, could continue to work.

Such a splitting out would also make it easier to introduce pull requests for very platform specific features, and would lessen the burden on those maintaining libc to know all of its oddities on many platforms.

@ssokolow Interesting. However, like communism, these sorts of centralizing approaches always seem to fail in practice (unless there's serious money to be made). It's a bit like the sirens; the call is alluring and impossible to counter, yet the reality is often ugly.

I was pondering on how best to tackle a newer libc structure, too, to make it easier to maintain and observe where things are 'missing'. Something that seems useful is to have a layout of code in files which mirrors the public headers. Perhaps headers could even map roughly to modules? There can't be a 1:1 mapping, and there would always be exceptions, but it might make some things a touch easier.

**Comment by gnzlbg** [comment]
FYI the [`llvm-sys` crate](https://github.com/tari/llvm-sys.rs) provides bindings to different incompatible LLVM version, and it does so by maintaining different branches for each LLVM version, and different major crate versions: e.g. `llvm-sys` versions 50.0.0, 60.0.0, 60.1.0, .. which correspond to LLVM versions 5.0.0, 6.0.0, 6.0.1, etc. The patch version number is used to release newer versions of the crate.

That is, users just include the bindings for the LLVM version that they want to target by setting e.g. `llvm-sys = 60` and then they get the latest bindings for that version. 

If somebody wants to develop a crate to target, e.g., different versions of some OS, that's an strategy that might be worth considering.

**Comment by ssokolow** [comment]
@raphaelcohn 

> However, like communism, these sorts of centralizing approaches always seem to fail in practice (unless there's serious money to be made). It's a bit like the sirens; the call is alluring and impossible to counter, yet the reality is often ugly.

I have no idea what you're talking about. What centralizing?

The idea proposed is that, since glibc sometimes lags behind on wrapping Linux syscalls, maybe the Linux kernel devs should provide a Linux kernel syscalls library to ensure that they're available in a timely manner. It'd be no more centralized than having the same people maintain both the kernel source and the kernel menuconfig to make sure they stay in sync.

**Comment by raphaelcohn** [comment]
@ssokolow Given the way you describe the Linux kernel syscalls library proposed, then I don't see anything about it that couldn't also be provided by a similar Rust wrapper. That said, a large part of the effort in libc's like musl is cleaning up the inconsistency in Linux's syscalls and associated data structures and occasional mismatch with POSIX APIs.

**Comment by inferiorhumanorgans** [comment]
Would it be possible to back up a bit and determine the ideal support matrix?  For example, should libc limit support to the actively supported versions of an operating system (e.g. FreeBSD 11 & 12, OpenBSD 6.3 & 6.5, DragonFlyBSD 5.4)?  If so it seems like while there would be some churn in the actual flags, the maintenance burden shouldn't be so high.  With such a narrow support matrix the following config flags could, in theory, work: `bsd`, `freebsd` (a.k.a. `freebsd12`), `freebsd11`, `freebsd12`.

In either case the top-level README seems a bit out of date in terms of what's actually being tested.

**Comment by asomers** [comment]
So far people have proposed 4 basic solutions to this problem, and they're all inadequate:

* Adding a build script would break cross-compiling.
* Changing target triples into target quadruples (eg `amd64-unknown-openbsd6.1`) would be way too much work.
* Adding backwards-compat feature flags (eg `freebsd-11-compat`) doesn't work for non-backwards compatible changes, such as OpenBSD's.
* Splitting libc into subcrates and keeping only the minimum functionality in libc itself isn't really a solution at all; it just kicks the can down the road.

In addition, none of those solutions even attempt to address the problem faced by rare or proprietary OS forks.  Right now if somebody makes a new OS fork, like Bitrig, TrueOS, or yet another shiny Android fork, the only way to get Rust support is to add a new `target_os` in Rust .  That's way too much work for forks that will only change a handful of syscalls.  Worse, adding a new `target_os` isn't even possible for proprietary forks like Isilon's OneOS.  The only way to solve the problem is to absolve rust-lang/libc from the responsibility of maintaining C library bindings for every single OS version in the world.  OS's need the power to vendor their own libc crate.  I propose a new way to do this, which would require a small change to Cargo but no changes to libc.  Since it's not a libc change per se, let's discuss it on the forum here: https://internals.rust-lang.org/t/pre-rfc-global-source-replacement-in-cargo-for-os-bindings/9383 .

**Comment by myfreeweb** [comment]
> Adding a build script would break cross-compiling

We could generate bindings on the fly when natively compiling, and use stored bindings (the ones that exist now) when cross-compiling maybe?

**Comment by gnzlbg** [comment]
> We could generate bindings on the fly when natively compiling, and use stored bindings (the ones that exist now) when cross-compiling maybe?

I think we can do this for `const`s, but if the "stored" bindings and the "native" bindings differ in type signatures, crates that compile against one might fail to compile against the other (because they use the wrong types, etc.). That is, crates targeting libc would need to be able to handle the "stored" vs "native" binding differences to be able to compile against both using `cfg()`s, and at that point we are back to "we need some sort of binding versioning and some way to `cfg` on that".

**Comment by raphaelcohn** [comment]
@asomers A mad, mad option 5: Build a libc in Rust, and use that as the default. It actually solves a very large amount of the cross-compiling problem, too. It's certainly do-able for Linux and the BSDs.

I've played around with all of the 4 basic solutions you've mentioned, and all are far from ideal. Once one adds in the various horribleness of different CPU architectures, it just gets plain crazy (eg PowerPC and MIPS on Linux) and maintenance is tough.

I do think one thing we should give up on is trying to have lots of `#[cfg]` options on structs and the like just so we define things in one place. I've tried doing similar things in my own code (eg wrapping some obscure socket options for Linux) and it's actually just too hard to maintain or understand. Instead, it's one of the few times in my career I'd actually advocate duplication. Each libc, for each OS, gets it own **set** of files, one per named header in the original libc if possible. I'd actually go further, and name the files by function name / struct and, if necessary, architecture, too, eg:-

* `musl/sys/utsname.utsname.rs`
* `musl/sys/utsname.uname.rs`
* `musl/sys/signal.SIGWAIT.mips.rs`
* `musl/sys/signal.SIGWAIT.x86_64.rs`
* `musl/sys/signal.SIGEMT.mips.rs`

Many will find this too fine a break down. I've come to find structuring definitions like this extremely useful. At a glance I can see that there isn't a definition of `SIGEMT` for `x86-64`. It allows for changes to be fine-grained, and easy to add files for a new architecture. And too show when a particular architecture or variant has an extra struct, flag or the like.

Personally, I'd be tempted to go further, and define:-

* `musl/1.24/sys/utsname.utsname.rs`
* `musl/1.24/sys/utsname.uname.rs`
* `musl/1.24/sys/signal.SIGWAIT.mips.rs`
* `musl/1.24/sys/signal.SIGWAIT.x86_64.rs`
* `musl/1.24/sys/signal.SIGEMT.mips.rs`.
* `openbsd/6.1/sys/...`

And then use the crate's default `features` to pull in the latest version when not otherwise specified.

**Comment by gnzlbg** [comment]
> I do think one thing we should give up on is trying to have lots of #[cfg] options on structs and the like just so we define things in one place. [...] Each libc, for each OS, gets it own set of files,

This is how `libc` is already implemented ? 

> Personally, I'd be tempted to go further, and define:-

Breaking changes across OS versions are small, duplicating all the code for each version seems overkill. 

> A mad, mad option 5: Build a libc in Rust, 

Go for it! You are not the first one to have this idea, and others have already started doing that. Maybe you can help them!

**Comment by raphaelcohn** [comment]
@gnzlbg Not in the slightest! It's an illogical mess attempting some deduplication with inconsistent naming. Not really anyone's fault - it's the result of Rust's successful growth - and it demonstrated just how hard it is to create a consistent-ish solution. (It's actually quite a common (and healthy) problem in open source projects with lots of contributors). I find it very hard to navigate and understand when checking for support of particular function or struct.

One does not really need to duplicate if one wishes - one can always use a symlink. Indeed, that is potentially actually quite a good option, as it logically shows that, say the struct `foobar` is the same in version `1.24` and `1.25`. However, my argument is that this problem is one of the very few - indeed almost only - time in my 20 year career I'd argue against de-duplication and DRY.

**Comment by gnzlbg** [comment]
>  Not in the slightest! It's an illogical mess attempting some deduplication with inconsistent naming. 

PRs welcome I guess. The amount of work required to clean this up (independently of how messy this currently is) is insignificant compared to the amount of work required to re-implement the libc's of all OSes that Rust supports in Rust.

> I find it very hard to navigate and understand when checking for support of particular function or struct.

There are docs for all architectures that rust supports online: https://rust-lang.github.io/libc/#platform-specific-documentation You can click on an arch, click on a struct, click on `[src]`, and every time I've used it, it brings me to the relevant source code.

**Comment by mati865** [comment]
@radupopescu you might want to check [relibc](https://gitlab.redox-os.org/redox-os/relibc).

**Comment by asomers** [comment]
@raphaelcohn please don't rewrite every OS"s C library in Rust.  That's basically what Go tried, and it hasn't gone smoothly.  It turns out there's some hard stuff in the C libraries.  Plus, it doesn't help libc's portability problem.  It just kicks the can down the road.  As I described in that forum post, I think the solution is more decentralization; libc wouldn't have to support every single version of every single OS if Cargo allowed OSes to provide their own libcs.

**Comment by gnzlbg** [comment]
> libc wouldn't have to support every single version of every single OS if Cargo allowed OSes to provide their own libcs.

I suppose this could work for "less severe" breaking changes, like changing the value of a const. But some FreeBSD breakage has been caused by, e.g., changing the type of a struct field. 

If the OS replaces libc with its own one containing the fix, then what? AFAICT libstd, and all other code using that API, won't compile because of a type mismatch, right?

**Comment by asomers** [comment]
Actually, changing the value of a const is impossible to do in a backwards-compatible way.  The FreeBSD breakage you mention is less severe, because it's all backwards compatible thanks to ELF symbol versioning.  The problem arises because Rust ignores the system headers, which is necessary for cross-compiling.

If FreeBSD 12 were to install its own forked libc as I suggested, then libstd and rustc would still work just fine.  The rustc built from ports would use the correct libc for FreeBSD 12.  And the rustc built from rustup would use the correct libc for FreeBSD 11, which would still link and run just fine, but would lack support for FreeBSD 12 features like longer mountpoints and 64-bit inodes.  Problems would only arise if you assume that, for example, `libc::stat` is identical to `libstd::stat`.  I doubt there's much if any code that does that.

**Comment by gnzlbg** [comment]
> Actually, changing the value of a const is impossible to do in a backwards-compatible way.

I meant the Rust `const`, that would be, e.g., a C macro (`#define FOO 3`), an `enum` discriminant, etc. Those do not appear in the ABI of the C library (no symbols), but they need to be exposed in Rust. Replacing `libc` with a different one should be able to update these things.

> FreeBSD 12...11,

I was actually referring to changes like this, with happened when updating libc to support FreeBSD 11 instead of FreeBSD 10 (EOL): https://github.com/rust-lang/libc/pull/1222

Note how the type of `struct stack_t::ss_sp` changed from `*mut c_char` to `*mut c_void`.  This did not broke anything because nobody was using it, but if the standard library were passing that struct field to a function taking a `*mut c_char`, then this change would have broken it. I don't understand how replacing libc with a different one that has the same backwards incompatible change could avoid breakage in similar cases to this one, were a type changes.

**Comment by myfreeweb** [comment]
> please don't rewrite every OS"s C library in Rust. That's basically what Go tried, and it hasn't gone smoothly. It turns out there's some hard stuff in the C libraries.

Also the libc **is** the stable and public API on most systems. (Systems that don't have a penguin logo, that is.)

And doing what Go does is a *massive* pain for less popular CPU architecture + OS combos.

> forked libc

hm — rustc already comes with a libc, but it's private and everyone is forced to redownload it with cargo. Switching everything to the libc distributed with rustc, such that it's just like libcore/libstd, will automatically achieve "forked libc".

**Comment by raphaelcohn** [comment]
>> PRs welcome I guess
Life is too short! It's either a complete re-working of the entire structure, with a proper focus on complete implementation of everything in libc for every major variant, or nothing...

>> every time I've used it, it brings me to the relevant source code
Which is often difficult to truly get an understanding of with a lot of cross platform `[#cfg]`. Using the file system to visually see high-level difference is often superior.

@asomers : It's not really that hard to write most of a libc - just an awful lot of work. If one is looking to only support pure-Rust, or nearly pure Rust implementations, or modern 4.0+ (say) Linux, then a stripped back libc-alike is quite do-able for Linux. That said, this is definitely a 'mad, mad' option. However, I've always been very doubtful that the wins are worthwhile versus the effort. There might be some small inlining gains, but, for the sort of network server software I tend to write, the only real gain is strip away the (very thin) layers before making a syscall - and I try to avoid syscalls.

@myfreeweb : agreed - it's the public, stable API on old-fashioned OSes (;-). And it's a pain in the neck, especially for cross-compilation.

**Comment by gnzlbg** [comment]
>  If one is looking to only support pure-Rust, or nearly pure Rust implementations, or modern 4.0+ (say) Linux, then a stripped back libc-alike is quite do-able for Linux.

This issue is about finding a solution to breaking changes in platform APIs, yet Linux is one of the few OSes that never break their API. 

I don't know what problem re-implementing libc in Rust for Linux would solve, but I don't see how it could solve the one being discussed here, since Linux does not have it AFAICT.

**Comment by raphaelcohn** [comment]
@gnzblg: Agreed; just re-implementing libc in Rust for Linux is not going to solve the problem being discussed here. I'm happy to stop discussing it - it's a distraction.

**Comment by gnzlbg** [comment]
> And the rustc built from rustup would use the correct libc for FreeBSD 11, which would still link and run just fine, but would lack support for FreeBSD 12 features like longer mountpoints and 64-bit inodes. 

The assumption that by targetting FreeBSD 11 libc is usable on FreeBSD 12 is incorrect.

Consider `freelocale`, which in FreeBSD 11 and currently in `libc` 0.2.53 returns an `int`, but which was changed in `FreeBSD 12` to return nothing - this is an ABI breaking change. Code calling `freelocale` would work on FreeBSD 11, but invoke undefined behavior on FreeBSD 12. 

There is probably a subset of the FreeBSD 11 and 12 API surface that can be exposed such that it works correctly on both platforms, but FreeBSD 11 is not it.

**Comment by asomers** [comment]
`freelocale` is an unusual case.  It didn't get a backwards-compatibility version when its return type changed because nobody could find any evidence of any code that actually used the return type.

https://svnweb.freebsd.org/base?view=revision&revision=303495

Even if some Rust code attempted to use its return value, I still don't think that would qualify as UB, because of the FFI interface.  UB would only come into play if the compiler knew that there was no initialized return value.  However, in this case the compiler believes that the return value is initialized, and believes that it exists where the C calling convention says that it should be.  On amd64, that means that libc's `freelocale` would return whatever happens to be in `RAX` at the time.

**Comment by gnzlbg** [comment]
I think that the original proposal of @semarie here (https://github.com/rust-lang/libc/issues/570#issuecomment-292780813) is the best solution to this problem. 

IIUC, essentially there would not be a `target_os = "freebsd"` and instead there would be `target_os = "freebsd11", `target_os = "freebsd12"`, etc. and these are from Rust perspective different OSes. To maintain backward compatibility with code using `target_os = "freebsd"` we can just keep the current `-..-freebsd` target and define it to mean "freebsd11", that is, libc would "just" need to write `cfg(any(target_os = "freebsd", target_os = "freebsd11")))` everywhere instead of just one of them. 

I'm not sure if this is what @alexcrichton had in mind here (https://github.com/rust-lang/libc/issues/570#issuecomment-292967129) when they said that this cannot be supported right now, but AFAICT this should "just work" without any changes to the compiler. It probably wouldn't even require an RFC.

---

If somebody wanted to go the RFC route, then maybe a `target_os_version = "11.0"`  that's only available for some OSes, and errors for OSes for which it is not available, sounds doable. We would still need multiple targets, e.g., `..-unknown-freebsd`, `..-unknown-freebsd11`, `..-unknown-freebsd12`, etc. and these would have `target_os_version` defined to potentially different numbers, and then for targets like `..-linux-gnu` `target_os_version` wouldn't be defined at all.

**Comment by semarie** [comment]
@gnzlbg an (unaccepted and unrejected) [RFC](https://github.com/rust-lang/rfcs/pull/2048) already exists for that since somehow long time now.

**Comment by asomers** [comment]
@gnzlbg @semarie's proposal would work well for OpenBSD, but at great cost to the Rust ecosystem.  Adding target_os's for every version of FreeBSD and NetBSD would have even more cost.  And it wouldn't solve the problem for rare and proprietary OSes.  This is why I proposed an alternate solution at https://internals.rust-lang.org/t/pre-rfc-global-source-replacement-in-cargo-for-os-bindings/9383 .  In a nutshell:
* Rust would work on any OS based on one of the supported ones, including weird forks like the Nintendo Switch.
* Rustup would only support one version of each OS.  For example, Rustup would only provide a toolchain for the latest OpenBSD.  Older versions of OpenBSD would be supported through OpenBSD's package manager.
* Where possible, Rustup's toolchain would be backwards compatible with a range of OS versions.  For FreeBSD, that means the toolchain would target FreeBSD 11 (but could built crates targeting FreeBSD 12).  In a few years Rustup would retarget its toolchain to FreeBSD 12.
* lib would no longer have to worry about the differences between different platform versions.  In the case of FreeBSD, libc would target one particular version (probably 11), and a FreeBSD-12 specific libc would be provided by FreeBSD itself.
* Most of the code changes would be confined to Cargo.

**Comment by gnzlbg** [comment]
@asomers I remember your proposal and I think it is interesting as well. From my point of view, your proposal and @semarie 's both solve problems in the same domain, but they have quite different goals and constraints on the solution, which is what results in two very different approaches, both having different trade-offs, to the point that we actually could do both since they appear to me to be compatible with each other.

Maybe it would be worth it to have a mini-RFC here first about what goals and constraints do we have, and once we are on the same page about that, reconsider both solutions, and try to come up with something that satisfies them all.

**Comment by myfreeweb** [comment]
Hmm, looks like you've decided to check it in libc's `build.rs`, nothing needed from the compiler:

https://github.com/rust-lang/libc/blob/d5a599ece74cde441577db7ece34b59017ff7e6c/build.rs#L19-L23

but only for testing right now.

If you just do `<=` checks instead of specifically 11 and 12, and on `__FreeBSD_version` from `/usr/include/osreldate.h` instead of `freebsd-version` output, this could very well be the easiest solution for production as well! And in cross-compilation, just use 11 for max compatibility.

**Comment by gnzlbg** [comment]
@myfreeweb the current system is for private use within libc only. This allow us to implement FreeBSD12 APIs, keep track of the backwards incompatible changes, and enables testing whatever system we choose to use to support this.

**Comment by myfreeweb** [comment]
Yes. And I'm saying this can easily be turned into the actual solution that won't be just for private use!

**Comment by DemiMarie-parity** [comment]
> @gnzlbg @semarie's proposal would work well for OpenBSD, but at great cost to the Rust ecosystem. Adding target_os's for every version of FreeBSD and NetBSD would have even more cost. And it wouldn't solve the problem for rare and proprietary OSes. This is why I proposed an alternate solution at https://internals.rust-lang.org/t/pre-rfc-global-source-replacement-in-cargo-for-os-bindings/9383 . In a nutshell:
> 
>     * Rust would work on any OS based on one of the supported ones, including weird forks like the Nintendo Switch.
> 
>     * Rustup would only support one version of each OS.  For example, Rustup would only provide a toolchain for the latest OpenBSD.  Older versions of OpenBSD would be supported through OpenBSD's package manager.

OpenBSD doesn’t bump Rust versions in stable branches of the ports tree, so one needs to use OpenBSD-CURRENT if one wants a recent Rust on that platform.  This solution also does not work for cross-compilation.

For the open-source OSs, one option would be to run bindgen on `/usr/include` from the most recent release.  That is a lot of code, but one advantage is that it is very easy to automate, which makes maintenance simpler.

**Comment by semarie** [comment]
> OpenBSD doesn’t bump Rust versions in stable branches of the ports tree.
It is by design: stable port tree receive only security updates. But nothing prevent a user to build recent rust release on OpenBSD stable.

**Comment by DemiMarie-parity** [comment]
> > OpenBSD doesn’t bump Rust versions in stable branches of the ports tree.
> 
> It is by design: stable port tree receive only security updates. But nothing prevent a user to build recent rust release on OpenBSD stable.

Understandable.  That said, this prevents products that depend on the latest stable Rust (such as Firefox) from getting security updates.  Also, security vulnerabilities occasionally do appear in Rust itself.

**Comment by asomers** [comment]
This RFC proposes a good solution: https://github.com/rust-lang/rfcs/pull/3036 .

**Comment by 64** [comment]
@asomers How would that work exactly? Have code inside the `libc` crate which is protected by `#[cfg(min_libc_version = ...)]`?

**Comment by asomers** [comment]
> @asomers How would that work exactly? Have code inside the `libc` crate which is protected by `#[cfg(min_libc_version = ...)]`?

Exactly.  Then consumers can set the min_libc_version in their own .cargo/config files.

**Comment by RalfJung** [comment]
TIL that emscripten also likes to change their ABI a lot (see e.g. https://github.com/rust-lang/libc/pull/3962). Isn't that pretty much the same problem? I am a bit surprised that the discussion above seems to be entirely about the BSDs.
